timemachine/cpp/src/neighborlist.cu

```diff
@@ -220,59 +180,131 @@ void Neighborlist<RealType>::build_nblist_device(
+    }
+    if (*std::max_element(row_idxs.begin(), row_idxs.end()) >= N_) {
+        throw std::runtime_error("indices values must be less than N");
+    }
+    std::vector<unsigned int> column_indices = get_indices_difference<unsigned int>(N_, unique_idxs);
+    const size_t row_count = row_idxs.size();
+    const size_t col_count = column_indices.size();
+
+    DeviceBuffer<unsigned int> row_idx_buffer(row_count);
+    DeviceBuffer<unsigned int> column_idx_buffer(col_count);
+
+    row_idx_buffer.copy_from(&row_idxs[0]);
+
+    column_idx_buffer.copy_from(&column_indices[0]);
+
+    this->set_idxs_device(
+        col_count, row_count, column_idx_buffer.data, row_idx_buffer.data, static_cast<cudaStream_t>(0));
+    gpuErrchk(cudaDeviceSynchronize());
+}
+
+template <typename RealType> void Neighborlist<RealType>::reset_row_idxs() {
+    const cudaStream_t stream = static_cast<cudaStream_t>(0);
+    this->reset_row_idxs_device(stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
+}
+
+template <typename RealType> void Neighborlist<RealType>::reset_row_idxs_device(const cudaStream_t stream) {
+    const int tpb = warp_size;
+    const int blocks = ceil_divide(N_, tpb);
+    // Fill the indices with the 0 to N-1 indices, indicating 'normal' neighborlist operation
+    k_arange<<<blocks, tpb, 0, stream>>>(N_, d_column_idxs_);
+    gpuErrchk(cudaPeekAtLastError());
+    k_arange<<<blocks, tpb, 0, stream>>>(N_, d_row_idxs_);
+    gpuErrchk(cudaPeekAtLastError());
+    this->NR_ = N_;
+    this->NC_ = N_;
+}
+
+template <typename RealType> void Neighborlist<RealType>::resize(const int size) {
+    const cudaStream_t stream = static_cast<cudaStream_t>(0);
+    this->resize_device(size, stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
+}
+
+// Resize the Neighborlist to function on a different size. Note that this only allows
+// finding interactions on a smaller set of the system, will not increase the size of the underlying buffers.
+template <typename RealType> void Neighborlist<RealType>::resize_device(const int size, const cudaStream_t stream) {
+    if (size <= 0) {
+        throw std::runtime_error("size is must be at least 1");
+    }
+    if (size > max_size_) {
+        throw std::runtime_error("size is greater than max size");
+    }
+    this->N_ = size;
+    this->reset_row_idxs_device(stream);
+}
+
+template <typename RealType>
+void Neighborlist<RealType>::set_idxs_device(
+    const int NC, const int NR, unsigned int *column_idxs, unsigned int *row_idxs, const cudaStream_t stream) {
+    if (NC + NR != N_) {
+        throw std::runtime_error("Total of indices must equal N");
+    }
+    const size_t tpb = warp_size;
+    gpuErrchk(
+        cudaMemcpyAsync(d_column_idxs_, column_idxs, NC * sizeof(*d_column_idxs_), cudaMemcpyDeviceToDevice, stream));
+    gpuErrchk(cudaMemcpyAsync(d_row_idxs_, row_idxs, NR * sizeof(*d_row_idxs_), cudaMemcpyDeviceToDevice, stream));
```

Comment:
It is to allow construction of idxs entirely on the GPU which is important for local MD to be fast. Will add some comments

Reply:
Added comments here: [55bbd75](https://github.com/proteneer/timemachine/pull/740/commits/55bbd75dcade56f25a647553cc851c6417465e24)

Reply:
> 1. Should the arguments use the `d_` prefix convention to indicate device memory? Alternatively, could they be passed as `DeviceBuffer`s? (both would make it clear that we're passing in device buffers)

Originally had the `d_` prefix, and ended up with `d_row_idxs` and `d_row_idxs_` which I found more deceptive then relying on the convention of `_device` on the method. Passing in `DeviceBuffers` would break our convention of `_device` methods and in a future deboggling week would seem reasonable to switch to a broader use of DeviceBuffers for our device code. Will make the name more explicit for the input arrays: `d_input_row_idxs`.

> 2. I'm not sure I understand the reason for copying the device buffer contents, versus just updating pointers (e.g. `d_column_idxs_ = column_idxs`. If subtle, worth adding a comment? (since you also need to free the old buffer, maybe it's not actually any more efficient to do it this way than with a copy?)

If you did `d_column_idxs_ = column_idxs` the `set_row_idxs` function would cause segfaults when the device buffer that allocated the idxs leaves scope and is freed. It would also be changing the underlying buffers of a class, which in my mind breaks RAII though perhaps that is incorrect.